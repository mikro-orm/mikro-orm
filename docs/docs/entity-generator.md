---
title: Entity Generator
---

To generate entities from existing database schema, you can use the `EntityGenerator` helper. It lives in its own package called `@mikro-orm/entity-generator`:

```bash npm2yarn
npm install @mikro-orm/entity-generator
```

> The version needs to be aligned with the `@mikro-orm/core` package and the database driver package.

To use it, you need to register the `EntityGenerator` extension in your ORM config:

```ts
import { defineConfig } from '@mikro-orm/postgresql';
import { EntityGenerator } from '@mikro-orm/entity-generator';

export default defineConfig({
  dbName: 'test',
  extensions: [EntityGenerator],
});
```

Then you can use it either via CLI:

> To work with the CLI, first install `@mikro-orm/cli` package locally. The version needs to be aligned with the `@mikro-orm/core` package.

```sh
npx mikro-orm generate-entities --dump  # Dumps all generated entities
npx mikro-orm generate-entities --save --path=./my-entities  # Saves entities into given directory
```

Or you can create simple script where you initialize MikroORM like this:

```ts title="./generate-entities.ts"
import { MikroORM } from '@mikro-orm/core';
import { EntityGenerator } from '@mikro-orm/entity-generator';

(async () => {
  const orm = await MikroORM.init({
    discovery: {
      // we need to disable validation for no entities
      warnWhenNoEntities: false,
      extensions: [EntityGenerator],
    },
    dbName: 'your-db-name',
    // ...
  });
  const dump = await orm.entityGenerator.generate({
    save: true,
    path: process.cwd() + '/my-entities',
  });
  console.log(dump);
  await orm.close(true);
})();
```

Then run this script via `ts-node` (or compile it to plain JS and use `node`):

```sh
$ ts-node generate-entities
```

## Advanced configuration

The behaviour of the entity generator can be adjusted either via `entityGenerator` section in the ORM config, or with the `GenerateOptions` object (parameter of the `generate` method), which takes precedence over the global configuration.

Available options:

| Option                                                         | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
|----------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `schema: string`                                               | The target schema to generate entities for. Defaults to what the main config would use.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| `takeTables: (string \| RegExp)[]`                                         | Consider some database tables. Accepts array of table names and `RegExp`s. If there is a foreign key reference to a non-taken table, the generated code will be as if that foreign key did not exist.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| `skipTables: (string \| RegExp)[]`                                         | Ignore some database tables. Accepts array of table names and `RegExp`s. If there is a foreign key reference to a skipped table, the generated code will be as if that foreign key did not exist.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
| `skipColumns`                                                  | Ignore some database tables columns. Accepts an object, where keys are table names with schema prefix if available, values are arrays of column names and `RegExp`s. If a skipped column is the target of a foreign key reference, the generated code will look as if that foreign key did not exist.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| `save: boolean`                                                | Whether to save the generated entities as files.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
| `path: string`                                                 | Folder to save the generated entities in, if saving is enabled. Defaults to a folder called `generated-entities` inside the `baseDir` of the main config.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| `fileName: (className: string) => string`                      | Callback to override the entity file name. Defaults to the entity name.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      |
| `bidirectionalRelations: boolean`                              | By default, the `EntityGenerator` generates only owning sides of relations (e.g. M:1). If set to true, generates also the inverse sides for them                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             |
| `identifiedReferences: boolean`                                | If set to `true`, generate M:1 and 1:1 relations as wrapped references. Also wraps any `lazy` properties.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| `entitySchema: boolean`                                        | By default, generate entities using decorators. If `true`, generate the entities using `EntitySchema` instead.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               |
| `esmImport: boolean`                                           | By default, import statements for entities without extensions are used. If set to `true`, uses ESM style import for imported entities, i.e. adds a `.js` suffix as extension.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| `scalarTypeInDecorator: boolean`                               | If `true`, include the `type` option in scalar property decorators. This information is discovered at runtime, but the process of discovery can be skipped by including this option in the decorator. If using `EntitySchema`, this type information is always included.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
| `scalarPropertiesForRelations: 'never' \| 'always' \| 'smart'` | <ul><li> `'never'` (default) - Do not generate any scalar properties for columns covered by foreign key relations. This effectively forces the application to always provide the entire relation, or (if all columns in the relation are nullable) omit the entire relation.</li><li> `'always'` - Generate all scalar properties for all columns covered by foreign key relations. This enables the application to deal with code that disables foreign key checks.</li><li> `'smart'` - Only generate scalar properties for foreign key relations, where doing so is necessary to enable the management of rows where a composite foreign key is not enforced due to some columns being set to NULL. This enables the application to deal with all rows that could possibly be part of a table, even when foreign key checks are always enabled.</li></ul> |
| `onlyPurePivotTables: boolean`                                 | By default, M:N relations are allowed to use pivot tables containing additional columns. If set to `true`, M:N relations will not be generated for such pivot tables.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
| `readOnlyPivotTables: boolean`                                 | By default, M:N relations are only generated if the collection would be writable, i.e. any additional columns need to be optional and have non-unique default values. If set to `true`, also generate M:N relations even if the collection would be read only (meaning the only way to write to it is by using the pivot entity directly). Such collections will include the `persist: false` option. This setting is effectively meaningless if `onlyPurePivotTables` is set to `true`.                                                                                                                                                                                                                                                                                                                                                                     |
| `customBaseEntityName: string`                                 | By default, entity classes do not extend a base class. If this option is set, a class with this name will be created, if it doesn't exist already, and any entity that does not inherit from another will be made to use this one. The class itself will also be set to contain a `[Config]` symbol with the currently running configuration.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
| `useCoreBaseEntity: boolean`                                   | By default, entity classes do not extend a base class. If this option is set to `true`, entity classes that do not extend another will be set to use the `BaseEntity` class in `@mikro-orm/core`. When combined with the `customBaseEntityName` option, effectively, only the generated base class will use the core's `BaseEntity` class.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
| `onInitialMetadata: MetadataProcessor`                         | A function can be provided to do processing of the initially collected metadata collection. Triggered after checking for tables and columns affected by skipping, but before base class generation, ManyToMany relation detection, and bi-directional relation detections. This hook is the appropriate place to add virtual and embeddable entities, and any owning sides of OneToOne or ManyToOne references. The generator can then figure out the inverse side and/or if any ManyToMany relations are affected. The function may return a promise, in which case, the promise will be awaited before moving on.                                                                                                                                                                                                                                          |
| `onProcessedMetadata: MetadataProcessor`                       | A function can be provided to do processing of the final metadata collection. Triggered after base class generation, ManyToMany relation detection, and bi-directional relation detection, just before outputting the metadata into files. This is the appropriate place for changes that need to be universally applied, including to ManyToMany and inverse side relation properties. The function may return a promise, in which case, the promise will be awaited before moving on.                                                                                                                                                                                                                                                                                                                                                                      |

Example configuration:

```ts
const dump = await orm.entityGenerator.generate({
  entitySchema: true,
  bidirectionalRelations: true,
  identifiedReferences: true,
  esmImport: true,
  save: true,
  path: process.cwd() + '/my-entities',
  skipTables: ['book', 'author'],
  skipColumns: {
    'public.user': ['email', 'middle_name'],
  },
});

```

## Processing of generated metadata

There are several features that are impossible to infer from the database schema alone.

The configuration settings `onInitialMetadata` and `onProcessedMetadata` can be used to add that information to the generated metadata, before the metadata is ultimately saved as a file. Note that this is different from the `onMetadata` hook in the configuration. The `onMetadata` hook does not affect the entity files, but those in the generator options do.

Some of the things that are suitable for these hooks include

- Adding the `hidden` flag on certain columns and relations.
- Adding the `lazy` flag on certain columns.
- Adding the `eager` flag on certain relations.
- Adding the `mapToPk` flag on certain relations.
- Adding the `orphanRemoval` flag on certain inverse relations.
- Adding the `cascade` option on certain relations.
- Tweaking the `type` of JSON columns to a specific (inline defined) shape.
- Adding Single Table Inheritance definitions.
- Adding `@Embedded` entities and references to them.
- Adding `virtual` entities.
- Adding properties with `formula`.

Here's an example that will make any column named "password" (regardless of what table is defined in) be lazy and hidden in its `onInitialMetadata` and in `onProcessedMetadata` will make all ManyToMany relations be hidden.

```ts
import { ReferenceKind } from '@mikro-orm/core';

await orm.entityGenerator.generate({
  onInitialMetadata: (metadata, platform) => {
    metadata.forEach(meta => {
      meta.props.forEach(prop => {
        if (prop.name === 'password') {
          prop.hidden = true;
          prop.lazy = true;
        }
      });
    });
  },
  onProcessedMetadata: (metadata, platform) => {
    metadata.forEach(meta => {
      meta.props.forEach(prop => {
        if (prop.kind === ReferenceKind.MANY_TO_MANY) {
          prop.hidden = true;
        }
      });
    });
  },
});
```

Adding embedded and virtual entities via the hooks can be somewhat tricky. The metadata objects are internal, and thus fewer checks are in place on them.

Here's an example that defines a simple embeddable with just two properties and adds a reference to it that is otherwise known to be a JSON column.

```ts
import { ReferenceKind } from '@mikro-orm/core';

await orm.entityGenerator.generate({
  onInitialMetadata: (metadata, platform) => {
    const embeddableEntityMeta = new EntityMetadata({
      className: 'IdentitiesContainer',
      collection: platform.getConfig().getNamingStrategy().classToTableName('IdentitiesContainer'),
      embeddable: true,
    });
    embeddableEntityMeta.addProperty({
      name: 'github',
      type: 'string',
      nullable: true,
      fieldNames: ['github'],
      columnTypes: ['varchar(255)'],
    });
    embeddableEntityMeta.addProperty({
      name: 'local',
      type: 'number',
      nullable: true,
      fieldNames: ['local'],
      columnTypes: ['int'],
    });
    metadata.push(embeddableEntityMeta);

    const identitiesPropOnAuthor = metadata.find(meta => meta.className === 'Author')!.properties.identities;
    identitiesPropOnAuthor.kind = ReferenceKind.EMBEDDED;
    identitiesPropOnAuthor.object = true;
    identitiesPropOnAuthor.type = 'IdentitiesContainer';
  },
});
```

An alternative approach is to define the embedded entities manually. The generator's metadata hooks only need to add the references to these embedded entities.

## Current limitations

- In MySQL, `TINYINT(1)` columns will be defined as boolean properties. There is no true `BOOLEAN` type in MySQL (the keyword is just an alias for `TINYINT(1)`).
- MongoDB is not supported.
