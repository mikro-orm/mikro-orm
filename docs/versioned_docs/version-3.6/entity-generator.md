---
title: Entity Generator
---

To generate entities from existing database schema, you can use `EntityGenerator` helper.

You can use it via CLI:

```sh
npx mikro-orm generate-entities --dump  # Dumps all generated entities
npx mikro-orm generate-entities --save --path=./my-entities  # Saves entities into given directory
```

Or you can create simple script where you initialize MikroORM like this:

```typescript title="./generate-entities.ts"
import { MikroORM } from 'mikro-orm';

(async () => {
  const orm = await MikroORM.init({
    dbName: 'your-db-name',
    // ...
  });
  const generator = orm.getEntityGenerator();
  const dump = await generator.generate({
    save: true,
    baseDir: process.cwd() + '/my-entities',
  });
  console.log(dump);
  await orm.close(true);
})();
```

Then run this script via `ts-node` (or compile it to plain JS and use `node`):

```sh
$ ts-node generate-entities
```

## Current limitations

- many to many relations are not supported, pivot table will be represented as separate entity
- in mysql, tinyint columns will be defined as boolean properties
