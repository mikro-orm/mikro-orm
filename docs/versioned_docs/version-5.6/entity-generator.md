---
title: Entity Generator
---

To generate entities from existing database schema, you can use `EntityGenerator` helper.

You can use it via CLI:

> To work with the CLI, first install `@mikro-orm/cli` package locally. The version needs to be aligned with the `@mikro-orm/core` package.

```sh
npx mikro-orm generate-entities --dump  # Dumps all generated entities
npx mikro-orm generate-entities --save --path=./my-entities  # Saves entities into given directory
```

Or you can create simple script where you initialize MikroORM like this:

```ts title="./generate-entities.ts"
import { MikroORM } from '@mikro-orm/core';

(async () => {
  const orm = await MikroORM.init({
    discovery: {
      // we need to disable validation for no entities
      warnWhenNoEntities: false,
    },
    dbName: 'your-db-name',
    // ...
  });
  const generator = orm.getEntityGenerator();
  const dump = await generator.generate({
    save: true,
    baseDir: process.cwd() + '/my-entities',
  });
  console.log(dump);
  await orm.close(true);
})();
```

Then run this script via `ts-node` (or compile it to plain JS and use `node`):

```sh
$ ts-node generate-entities
```

## Advanced configuration

By default, the `EntityGenerator` generates only owning sides of relations (e.g. M:1) and uses decorators for the entity definition. We can adjust its behaviour via `entityGenerator` section in the ORM config. Available options:

- `bidirectionalRelations` to generate also the inverse sides for them
- `identifiedReferences` to generate M:1 and 1:1 relations as wrapped references
- `entitySchema` to generate the entities using `EntitySchema` instead of decorators
- `esmImport` to use esm style import for imported entities e.x. when `esmImport=true`, generated entities include `import Author from './Author.js'`

## Current limitations

- in mysql, tinyint columns will be defined as boolean properties
