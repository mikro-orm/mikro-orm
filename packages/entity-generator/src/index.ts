/**
 * @packageDocumentation
 * @module entity-generator
 */
export * from './EntityGenerator.js';
