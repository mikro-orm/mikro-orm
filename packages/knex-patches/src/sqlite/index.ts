export * from './SqliteKnexDialect';
export * from './SqliteTableCompiler';
