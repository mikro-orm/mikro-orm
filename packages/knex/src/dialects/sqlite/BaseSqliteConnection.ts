import type { Knex } from 'knex';
import { ensureDir, readFile } from 'fs-extra';
import { dirname } from 'path';
import { AbstractSqlConnection } from '../../AbstractSqlConnection';
import { Utils } from '@mikro-orm/core';

export abstract class BaseSqliteConnection extends AbstractSqlConnection {

  override async connect(): Promise<void> {
    this.createKnex();
    await ensureDir(dirname(this.config.get('dbName')!));
    await this.client.raw('pragma foreign_keys = on');
  }

  getDefaultClientUrl(): string {
    return '';
  }

  override getClientUrl(): string {
    return '';
  }

  override async loadFile(path: string): Promise<void> {
    const conn = await this.client.client.acquireConnection();
    await conn.exec((await readFile(path)).toString());
    await this.client.client.releaseConnection(conn);
  }

  protected override getKnexOptions(type: string): Knex.Config {
    return Utils.mergeConfig({
      client: type,
      connection: {
        filename: this.config.get('dbName'),
      },
      pool: this.config.get('pool'),
      useNullAsDefault: true,
    }, this.config.get('driverOptions'));
  }

  protected transformRawResult<T>(res: any, method: 'all' | 'get' | 'run'): T {
    if (method === 'get') {
      return res[0];
    }

    if (method === 'all') {
      return res;
    }

    if (Array.isArray(res)) {
      return {
        insertId: res[res.length - 1]?.id ?? 0,
        affectedRows: res.length,
        row: res[0],
        rows: res,
      } as T;
    }

    return {
      insertId: res.lastInsertRowid,
      affectedRows: res.changes,
    } as unknown as T;
  }

}
