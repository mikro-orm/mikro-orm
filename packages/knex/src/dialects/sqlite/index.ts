export * from './BaseSqliteConnection';
export * from './BaseSqlitePlatform';
export * from './BaseSqliteSchemaHelper';
export * from './SqliteTableCompiler';
export * from './BetterSqliteKnexDialect';
export * from './LibSqlKnexDialect';
export * from './SqliteKnexDialect';
