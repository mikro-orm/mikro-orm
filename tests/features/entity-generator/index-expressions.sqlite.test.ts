import { MikroORM } from '@mikro-orm/sqlite';
import { EntityGenerator } from '@mikro-orm/entity-generator';

test('4911', async () => {
  const orm = await MikroORM.init({
    dbName: ':memory:',
    discovery: {
      warnWhenNoEntities: false,
    },
    extensions: [EntityGenerator],
  });
  await orm.schema.execute(`
CREATE TABLE "dcim_device" ("created" timestamptz, "id" int8 NOT NULL, "name" varchar, "_name" varchar, "asset_tag" varchar, "position" numeric, "face" varchar NOT NULL, "vc_position" int2 CHECK (vc_position >= 0), "vc_priority" int2 CHECK (vc_priority >= 0), "cluster_id" int8, "role_id" int8 NOT NULL, "device_type_id" int8 NOT NULL, "location_id" int8, "platform_id" int8, "primary_ip4_id" int8, "primary_ip6_id" int8, "rack_id" int8, "site_id" int8 NOT NULL, "tenant_id" int8, "virtual_chassis_id" int8, "airflow" varchar NOT NULL, "config_template_id" int8, "oob_ip_id" int8, PRIMARY KEY ("id"));
CREATE INDEX dcim_device_virtual_chassis_id_aed51693 ON "dcim_device" (virtual_chassis_id);
CREATE UNIQUE INDEX dcim_device_unique_virtual_chassis_vc_position ON "dcim_device" (virtual_chassis_id, vc_position);
CREATE UNIQUE INDEX dcim_device_unique_rack_position_face ON "dcim_device" (rack_id, position, face);
CREATE UNIQUE INDEX dcim_device_unique_name_site_tenant ON "dcim_device" (lower(name), site_id, tenant_id);
CREATE UNIQUE INDEX dcim_device_unique_name_site ON "dcim_device" (lower(name), site_id) WHERE tenant_id IS NULL;
CREATE INDEX dcim_device_tenant_id_dcea7969 ON "dcim_device" (tenant_id);
CREATE INDEX dcim_device_site_id_ff897cf6 ON "dcim_device" (site_id);
CREATE INDEX dcim_device_rack_id_23bde71f ON "dcim_device" (rack_id);
CREATE UNIQUE INDEX dcim_device_primary_ip6_id_key ON "dcim_device" (primary_ip6_id);
CREATE UNIQUE INDEX dcim_device_primary_ip4_id_key ON "dcim_device" (primary_ip4_id);
CREATE INDEX dcim_device_platform_id_468138f1 ON "dcim_device" (platform_id);
CREATE INDEX dcim_device_oob_ip_id_key ON "dcim_device" (oob_ip_id);
CREATE INDEX dcim_device_location_id_11a7bedb ON "dcim_device" (location_id);
CREATE INDEX dcim_device_device_type_id_d61b4086 ON "dcim_device" (device_type_id);
CREATE INDEX dcim_device_device_role_id_682e8188 ON "dcim_device" (role_id);
CREATE INDEX dcim_device_config_template_id_316328c4 ON "dcim_device" (config_template_id);
CREATE INDEX dcim_device_cluster_id_cf852f78 ON "dcim_device" (cluster_id);
CREATE UNIQUE INDEX dcim_device_asset_tag_key ON "dcim_device" (asset_tag);
CREATE INDEX dcim_device_asset_tag_8dac1079_like ON "dcim_device" (asset_tag);
  `);
  const dump = await orm.entityGenerator.generate();
  expect(dump).toMatchSnapshot();
  await orm.close(true);
});
