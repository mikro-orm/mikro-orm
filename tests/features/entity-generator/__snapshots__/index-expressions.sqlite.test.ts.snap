// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`4911 1`] = `
[
  "import { Entity, Index, PrimaryKey, Property, Unique } from '@mikro-orm/core';

@Entity()
@Unique({ name: 'dcim_device_unique_rack_position_face', properties: ['rackId', 'position', 'face'] })
@Unique({ name: 'dcim_device_unique_virtual_chassis_vc_position', properties: ['virtualChassisId', 'vcPosition'] })
export class DcimDevice {

  @Property({ columnType: 'timestamptz', nullable: true })
  created?: unknown;

  @PrimaryKey({ columnType: 'int8' })
  id!: unknown;

  @Property({ nullable: true })
  name?: string;

  @Property({ nullable: true })
  _name?: string;

  @Index({ name: 'dcim_device_asset_tag_8dac1079_like' })
  @Unique({ name: 'dcim_device_asset_tag_key' })
  @Property({ nullable: true })
  assetTag?: string;

  @Property({ type: 'decimal', nullable: true })
  position?: string;

  @Property()
  face!: string;

  @Property({ columnType: 'int2', nullable: true })
  vcPosition?: unknown;

  @Property({ columnType: 'int2', nullable: true })
  vcPriority?: unknown;

  @Index({ name: 'dcim_device_cluster_id_cf852f78' })
  @Property({ columnType: 'int8', nullable: true })
  clusterId?: unknown;

  @Index({ name: 'dcim_device_device_role_id_682e8188' })
  @Property({ columnType: 'int8' })
  roleId!: unknown;

  @Index({ name: 'dcim_device_device_type_id_d61b4086' })
  @Property({ columnType: 'int8' })
  deviceTypeId!: unknown;

  @Index({ name: 'dcim_device_location_id_11a7bedb' })
  @Property({ columnType: 'int8', nullable: true })
  locationId?: unknown;

  @Index({ name: 'dcim_device_platform_id_468138f1' })
  @Property({ columnType: 'int8', nullable: true })
  platformId?: unknown;

  @Unique({ name: 'dcim_device_primary_ip4_id_key' })
  @Property({ fieldName: 'primary_ip4_id', columnType: 'int8', nullable: true })
  primaryIp4Id?: unknown;

  @Unique({ name: 'dcim_device_primary_ip6_id_key' })
  @Property({ fieldName: 'primary_ip6_id', columnType: 'int8', nullable: true })
  primaryIp6Id?: unknown;

  @Index({ name: 'dcim_device_rack_id_23bde71f' })
  @Property({ columnType: 'int8', nullable: true })
  rackId?: unknown;

  @Index({ name: 'dcim_device_site_id_ff897cf6' })
  @Property({ columnType: 'int8' })
  siteId!: unknown;

  @Index({ name: 'dcim_device_tenant_id_dcea7969' })
  @Property({ columnType: 'int8', nullable: true })
  tenantId?: unknown;

  @Index({ name: 'dcim_device_virtual_chassis_id_aed51693' })
  @Property({ columnType: 'int8', nullable: true })
  virtualChassisId?: unknown;

  @Property()
  airflow!: string;

  @Index({ name: 'dcim_device_config_template_id_316328c4' })
  @Property({ columnType: 'int8', nullable: true })
  configTemplateId?: unknown;

  @Index({ name: 'dcim_device_oob_ip_id_key' })
  @Property({ columnType: 'int8', nullable: true })
  oobIpId?: unknown;

}
",
]
`;
